File: backend/services/storage.go
package services

import (
	"context"
	"fmt"
	"io"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
)

// ObjectInfo is a slim projection of S3 ListObjectsV2 entries used by paged
// walkers. Carrying Size + LastModified inline avoids a per-key HEAD round
// trip in the orphan-sweep age check.
type ObjectInfo struct {
	Key          string
	Size         int64
	LastModified time.Time
}

type StorageService struct {
	client *s3.Client
	bucket string
}

func NewStorage(endpoint, accessKey, secretKey, bucket, region string) (*StorageService, error) {
	cfg, err := awsconfig.LoadDefaultConfig(context.Background(),
		awsconfig.WithRegion(region),
		awsconfig.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(accessKey, secretKey, "")),
	)
	if err != nil {
		return nil, fmt.Errorf("aws config: %w", err)
	}

	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.BaseEndpoint = aws.String(endpoint)
		o.UsePathStyle = true // SeaweedFS requires path-style
	})

	return &StorageService{client: client, bucket: bucket}, nil
}

// Upload streams data directly to SeaweedFS — no disk buffering.
func (s *StorageService) Upload(ctx context.Context, path string, body io.Reader, size int64) error {
	_, err := s.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:        aws.String(s.bucket),
		Key:           aws.String(path),
		Body:          body,
		ContentLength: aws.Int64(size),
	})
	if err != nil {
		return fmt.Errorf("s3 put: %w", err)
	}
	return nil
}

// PutObjectVersioned puts an object and returns the SeaweedFS version id.
// Bucket must have versioning enabled (see seaweedfs-init.sh / lifecycle.json).
func (s *StorageService) PutObjectVersioned(ctx context.Context, key string, body io.Reader, size int64) (string, error) {
	out, err := s.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:        aws.String(s.bucket),
		Key:           aws.String(key),
		Body:          body,
		ContentLength: aws.Int64(size),
	})
	if err != nil {
		return "", err
	}
	if out.VersionId == nil {
		return "", nil // bucket might not have versioning enabled
	}
	return *out.VersionId, nil
}

// PresignedDownload generates a presigned URL valid for 15 minutes.
func (s *StorageService) PresignedDownload(ctx context.Context, path string) (string, error) {
	presigner := s3.NewPresignClient(s.client)
	req, err := presigner.PresignGetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(path),
	}, s3.WithPresignExpires(15*time.Minute))
	if err != nil {
		return "", fmt.Errorf("presign: %w", err)
	}
	return req.URL, nil
}

// GetObject fetches an object from SeaweedFS and returns its body and size.
func (s *StorageService) GetObject(ctx context.Context, path string) (io.ReadCloser, int64, error) {
	result, err := s.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(path),
	})
	if err != nil {
		return nil, 0, fmt.Errorf("s3 get: %w", err)
	}
	size := int64(0)
	if result.ContentLength != nil {
		size = *result.ContentLength
	}
	return result.Body, size, nil
}

// GetObjectVersion fetches a specific S3 (SeaweedFS) noncurrent version of an object.
// Returns the body stream + content-length.
func (s *StorageService) GetObjectVersion(ctx context.Context, path, versionID string) (io.ReadCloser, int64, error) {
	result, err := s.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket:    aws.String(s.bucket),
		Key:       aws.String(path),
		VersionId: aws.String(versionID),
	})
	if err != nil {
		return nil, 0, fmt.Errorf("s3 get version: %w", err)
	}
	size := int64(0)
	if result.ContentLength != nil {
		size = *result.ContentLength
	}
	return result.Body, size, nil
}

// Delete removes an object from SeaweedFS.
func (s *StorageService) Delete(ctx context.Context, path string) error {
	_, err := s.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(path),
	})
	return err
}

// DeleteObjectVersion deletes a specific S3 (SeaweedFS) noncurrent version of an object.
func (s *StorageService) DeleteObjectVersion(ctx context.Context, key, versionID string) error {
	_, err := s.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket:    aws.String(s.bucket),
		Key:       aws.String(key),
		VersionId: aws.String(versionID),
	})
	return err
}

// ListObjectsPaged iterates ListObjectsV2 results in pages of up to 1000
// keys, invoking page() once per non-empty page. Stops + returns the
// error if page() returns one. Used by both DeletePrefix and the orphan
// sweep — extracting it lets both share one tested LIST loop.
func (s *StorageService) ListObjectsPaged(
	ctx context.Context,
	prefix string,
	page func(objs []ObjectInfo) error,
) error {
	var continuationToken *string
	for {
		out, err := s.client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
			Bucket:            aws.String(s.bucket),
			Prefix:            aws.String(prefix),
			ContinuationToken: continuationToken,
		})
		if err != nil {
			return fmt.Errorf("s3 list: %w", err)
		}
		if len(out.Contents) > 0 {
			objs := make([]ObjectInfo, 0, len(out.Contents))
			for _, obj := range out.Contents {
				if obj.Key == nil {
					continue
				}
				size := int64(0)
				if obj.Size != nil {
					size = *obj.Size
				}
				lm := time.Time{}
				if obj.LastModified != nil {
					lm = *obj.LastModified
				}
				objs = append(objs, ObjectInfo{Key: *obj.Key, Size: size, LastModified: lm})
			}
			if err := page(objs); err != nil {
				return err
			}
		}
		if out.IsTruncated == nil || !*out.IsTruncated {
			return nil
		}
		continuationToken = out.NextContinuationToken
	}
}

// DeleteObjectsBatch removes up to 1000 keys in a single DeleteObjects S3 call.
// S3 caps a delete batch at exactly 1000 keys; callers should chunk if larger.
func (s *StorageService) DeleteObjectsBatch(ctx context.Context, keys []string) error {
	if len(keys) == 0 {
		return nil
	}
	ids := make([]s3types.ObjectIdentifier, len(keys))
	for i := range keys {
		k := keys[i]
		ids[i] = s3types.ObjectIdentifier{Key: &k}
	}
	_, err := s.client.DeleteObjects(ctx, &s3.DeleteObjectsInput{
		Bucket: aws.String(s.bucket),
		Delete: &s3types.Delete{Objects: ids, Quiet: aws.Bool(true)},
	})
	if err != nil {
		return fmt.Errorf("s3 delete batch: %w", err)
	}
	return nil
}

// DeletePrefix wipes every object whose key begins with the given prefix.
// Used to GC all per-file children (snapshot blob + asset blobs + …) when
// the parent file is deleted from Drive.
//
// Returns the first error encountered. Callers treat this as best-effort —
// the parent file row is already gone from the DB, so a partial failure
// only leaks orphan blobs (recoverable later by the admin orphan sweep).
func (s *StorageService) DeletePrefix(ctx context.Context, prefix string) error {
	return s.ListObjectsPaged(ctx, prefix, func(objs []ObjectInfo) error {
		keys := make([]string, len(objs))
		for i, o := range objs {
			keys[i] = o.Key
		}
		return s.DeleteObjectsBatch(ctx, keys)
	})
}

// CompletedPart is the {PartNumber, ETag} pair S3 needs at finalize time.
// We model it ourselves so the tus DB row can serialise it as JSONB
// without dragging the SDK type into the schema.
type CompletedPart struct {
	PartNumber int32  `json:"part_number"`
	ETag       string `json:"etag"`
}

// CreateMultipart opens a new S3 multipart upload at `key`. Returns the
// opaque UploadId that subsequent UploadPart / CompleteMultipart calls
// must thread through. The lifetime of the multipart on SeaweedFS is
// bounded by the stale-upload sweeper — if the user abandons mid-stream
// we Abort it explicitly.
func (s *StorageService) CreateMultipart(ctx context.Context, key string) (string, error) {
	out, err := s.client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		return "", fmt.Errorf("s3 create multipart: %w", err)
	}
	if out.UploadId == nil {
		return "", fmt.Errorf("s3 create multipart: empty upload id")
	}
	return *out.UploadId, nil
}

// UploadPart streams one part of a multipart upload. partNumber is 1-based
// per S3. Returns the ETag the caller must remember for CompleteMultipart.
// Note: S3 requires every part except the last to be ≥ 5 MiB. The tus
// handler enforces that — clients sending sub-5MiB PATCHes before the
// final one will get a 4xx from S3 surfaced as 500 here.
func (s *StorageService) UploadPart(
	ctx context.Context,
	key, uploadID string,
	partNumber int32,
	body io.Reader,
	size int64,
) (string, error) {
	out, err := s.client.UploadPart(ctx, &s3.UploadPartInput{
		Bucket:        aws.String(s.bucket),
		Key:           aws.String(key),
		UploadId:      aws.String(uploadID),
		PartNumber:    aws.Int32(partNumber),
		Body:          body,
		ContentLength: aws.Int64(size),
	})
	if err != nil {
		return "", fmt.Errorf("s3 upload part %d: %w", partNumber, err)
	}
	if out.ETag == nil {
		return "", fmt.Errorf("s3 upload part %d: empty etag", partNumber)
	}
	return *out.ETag, nil
}

// CompleteMultipart finalises the multipart upload, producing one object
// at `key`. Parts must be passed in PartNumber order.
func (s *StorageService) CompleteMultipart(
	ctx context.Context,
	key, uploadID string,
	parts []CompletedPart,
) error {
	sdkParts := make([]s3types.CompletedPart, len(parts))
	for i, p := range parts {
		etag := p.ETag
		pn := p.PartNumber
		sdkParts[i] = s3types.CompletedPart{
			PartNumber: aws.Int32(pn),
			ETag:       aws.String(etag),
		}
	}
	_, err := s.client.CompleteMultipartUpload(ctx, &s3.CompleteMultipartUploadInput{
		Bucket:          aws.String(s.bucket),
		Key:             aws.String(key),
		UploadId:        aws.String(uploadID),
		MultipartUpload: &s3types.CompletedMultipartUpload{Parts: sdkParts},
	})
	if err != nil {
		return fmt.Errorf("s3 complete multipart: %w", err)
	}
	return nil
}

// AbortMultipart discards a multipart upload — used when the user cancels
// or when the stale-upload sweeper reaps an abandoned session. Idempotent
// per the S3 spec: aborting an already-aborted upload returns success.
func (s *StorageService) AbortMultipart(ctx context.Context, key, uploadID string) error {
	_, err := s.client.AbortMultipartUpload(ctx, &s3.AbortMultipartUploadInput{
		Bucket:   aws.String(s.bucket),
		Key:      aws.String(key),
		UploadId: aws.String(uploadID),
	})
	if err != nil {
		return fmt.Errorf("s3 abort multipart: %w", err)
	}
	return nil
}

// CopyObject moves an object within the bucket from src to dst. Used by
// the tus finaliser to rename the temp multipart-assembled object to its
// canonical {userId}/{collectionId}/{fileId} path, then DeleteObject the
// temp. (Naïve "rename" doesn't exist in S3 — CopyObject + Delete is the
// convention.)
func (s *StorageService) CopyObject(ctx context.Context, src, dst string) error {
	_, err := s.client.CopyObject(ctx, &s3.CopyObjectInput{
		Bucket:     aws.String(s.bucket),
		Key:        aws.String(dst),
		CopySource: aws.String(s.bucket + "/" + src),
	})
	if err != nil {
		return fmt.Errorf("s3 copy %s -> %s: %w", src, dst, err)
	}
	return nil
}

File: crates/kutup-server/src/storage.rs
//! S3 (SeaweedFS) storage service — mirrors `backend/services/storage.go`
//! (`aws-sdk-go-v2` → `aws-sdk-s3`).
//!
//! Path-style addressing + a static-credentials provider, exactly like the Go
//! `NewStorage`. Covers the object get/put/delete + prefix-wipe paths (files/versions/
//! assets), the multipart paths (tus), and presigned download (public shares). Go's
//! `CopyObject` is unported — it became dead after the tus temp→canonical-key change and
//! has no caller. Version-delete (cleanup) lands with slice 7.

use std::time::Duration;

use anyhow::{Context, Result};
use aws_sdk_s3::config::{BehaviorVersion, Credentials, Region};
use aws_sdk_s3::presigning::PresigningConfig;
use aws_sdk_s3::primitives::ByteStream;
use aws_sdk_s3::types::{
    CompletedMultipartUpload, CompletedPart as S3CompletedPart, Delete, ObjectIdentifier,
};
use aws_sdk_s3::Client;
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

/// One object's metadata from a LIST — mirrors `services.ObjectInfo`. Used by the orphan
/// sweep to age-filter candidates.
#[derive(Clone, Debug)]
pub struct ObjectInfo {
    pub key: String,
    pub size: i64,
    pub last_modified: OffsetDateTime,
}

/// The `{PartNumber, ETag}` pair S3 needs at finalize time — mirrors
/// `services.CompletedPart`. Serialised into the `uploads.s3_part_etags` JSONB column;
/// the snake_case field names match the Go `json:"part_number"`/`json:"etag"` tags so a
/// row written by either backend round-trips through the other.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CompletedPart {
    pub part_number: i32,
    pub etag: String,
}

/// Wraps the S3 client + target bucket — mirrors `StorageService`.
#[derive(Clone)]
pub struct StorageService {
    client: Client,
    bucket: String,
}

impl StorageService {
    /// Builds the client with path-style addressing + static creds — mirrors `NewStorage`.
    pub fn new(
        endpoint: &str,
        access_key: &str,
        secret_key: &str,
        bucket: &str,
        region: &str,
    ) -> Self {
        let creds = Credentials::new(access_key, secret_key, None, None, "kutup-static");
        let conf = aws_sdk_s3::config::Builder::new()
            .behavior_version(BehaviorVersion::latest())
            .region(Region::new(region.to_string()))
            .credentials_provider(creds)
            .endpoint_url(endpoint)
            .force_path_style(true) // SeaweedFS requires path-style
            .build();
        StorageService {
            client: Client::from_conf(conf),
            bucket: bucket.to_string(),
        }
    }

    /// Streams data to S3 — mirrors `Upload`.
    pub async fn upload(&self, path: &str, body: ByteStream, size: i64) -> Result<()> {
        self.client
            .put_object()
            .bucket(&self.bucket)
            .key(path)
            .body(body)
            .content_length(size)
            .send()
            .await
            .context("s3 put")?;
        Ok(())
    }

    /// Puts an object and returns the SeaweedFS version id (empty if unversioned) —
    /// mirrors `PutObjectVersioned`.
    pub async fn put_object_versioned(
        &self,
        key: &str,
        body: ByteStream,
        size: i64,
    ) -> Result<String> {
        let out = self
            .client
            .put_object()
            .bucket(&self.bucket)
            .key(key)
            .body(body)
            .content_length(size)
            .send()
            .await
            .context("s3 put versioned")?;
        Ok(out.version_id().unwrap_or("").to_string())
    }

    /// Generates a presigned GET URL valid for 15 minutes — mirrors `PresignedDownload`.
    /// Used by the public-share download endpoint.
    pub async fn presigned_download(&self, path: &str) -> Result<String> {
        let cfg =
            PresigningConfig::expires_in(Duration::from_secs(15 * 60)).context("presign config")?;
        let req = self
            .client
            .get_object()
            .bucket(&self.bucket)
            .key(path)
            .presigned(cfg)
            .await
            .context("presign")?;
        Ok(req.uri().to_string())
    }

    /// Fetches an object — mirrors `GetObject`. Returns the body stream + content length.
    pub async fn get_object(&self, path: &str) -> Result<(ByteStream, i64)> {
        let out = self
            .client
            .get_object()
            .bucket(&self.bucket)
            .key(path)
            .send()
            .await
            .context("s3 get")?;
        let size = out.content_length().unwrap_or(0);
        Ok((out.body, size))
    }

    /// Deletes a specific (noncurrent) object version — mirrors `DeleteObjectVersion`.
    /// Used by the version-cleanup job.
    pub async fn delete_object_version(&self, key: &str, version_id: &str) -> Result<()> {
        self.client
            .delete_object()
            .bucket(&self.bucket)
            .key(key)
            .version_id(version_id)
            .send()
            .await
            .context("s3 delete version")?;
        Ok(())
    }

    /// Lists one page (≤1000 keys) under `prefix`, returning the objects + the continuation
    /// token for the next page (`None` when exhausted) — the paged half of Go's
    /// `ListObjectsPaged`. The orphan sweep drives the loop so it can do per-page DB work +
    /// inter-page sleeps.
    pub async fn list_objects_page(
        &self,
        prefix: &str,
        token: Option<String>,
    ) -> Result<(Vec<ObjectInfo>, Option<String>)> {
        let out = self
            .client
            .list_objects_v2()
            .bucket(&self.bucket)
            .prefix(prefix)
            .set_continuation_token(token)
            .send()
            .await
            .context("s3 list")?;
        let objs = out
            .contents()
            .iter()
            .filter_map(|o| {
                let key = o.key()?.to_string();
                let size = o.size().unwrap_or(0);
                let last_modified = o
                    .last_modified()
                    .and_then(|d| OffsetDateTime::from_unix_timestamp(d.secs()).ok())
                    .unwrap_or(OffsetDateTime::UNIX_EPOCH);
                Some(ObjectInfo {
                    key,
                    size,
                    last_modified,
                })
            })
            .collect();
        let next = if out.is_truncated() == Some(true) {
            out.next_continuation_token().map(String::from)
        } else {
            None
        };
        Ok((objs, next))
    }

    /// Fetches a specific noncurrent version — mirrors `GetObjectVersion`.
    pub async fn get_object_version(
        &self,
        path: &str,
        version_id: &str,
    ) -> Result<(ByteStream, i64)> {
        let out = self
            .client
            .get_object()
            .bucket(&self.bucket)
            .key(path)
            .version_id(version_id)
            .send()
            .await
            .context("s3 get version")?;
        let size = out.content_length().unwrap_or(0);
        Ok((out.body, size))
    }

    /// Removes an object — mirrors `Delete`.
    pub async fn delete(&self, path: &str) -> Result<()> {
        self.client
            .delete_object()
            .bucket(&self.bucket)
            .key(path)
            .send()
            .await
            .context("s3 delete")?;
        Ok(())
    }

    /// Deletes up to 1000 keys in one call — mirrors `DeleteObjectsBatch`.
    pub async fn delete_objects_batch(&self, keys: &[String]) -> Result<()> {
        if keys.is_empty() {
            return Ok(());
        }
        let objects: Vec<ObjectIdentifier> = keys
            .iter()
            .map(|k| ObjectIdentifier::builder().key(k).build())
            .collect::<Result<_, _>>()
            .context("build delete identifiers")?;
        let delete = Delete::builder()
            .set_objects(Some(objects))
            .quiet(true)
            .build()
            .context("build delete")?;
        self.client
            .delete_objects()
            .bucket(&self.bucket)
            .delete(delete)
            .send()
            .await
            .context("s3 delete batch")?;
        Ok(())
    }

    /// Wipes every object under a prefix — mirrors `DeletePrefix` (+ the paged LIST loop).
    /// Best-effort: callers have already removed the DB rows, so a partial failure only
    /// leaks orphan blobs (recoverable by the admin orphan sweep).
    pub async fn delete_prefix(&self, prefix: &str) -> Result<()> {
        let mut continuation: Option<String> = None;
        loop {
            let out = self
                .client
                .list_objects_v2()
                .bucket(&self.bucket)
                .prefix(prefix)
                .set_continuation_token(continuation.clone())
                .send()
                .await
                .context("s3 list")?;
            let keys: Vec<String> = out
                .contents()
                .iter()
                .filter_map(|o| o.key().map(String::from))
                .collect();
            if !keys.is_empty() {
                self.delete_objects_batch(&keys).await?;
            }
            if out.is_truncated() != Some(true) {
                return Ok(());
            }
            continuation = out.next_continuation_token().map(String::from);
        }
    }

    /// Opens a new S3 multipart upload at `key`, returning the opaque UploadId —
    /// mirrors `CreateMultipart`.
    pub async fn create_multipart(&self, key: &str) -> Result<String> {
        let out = self
            .client
            .create_multipart_upload()
            .bucket(&self.bucket)
            .key(key)
            .send()
            .await
            .context("s3 create multipart")?;
        out.upload_id()
            .map(String::from)
            .context("s3 create multipart: empty upload id")
    }

    /// Streams one part of a multipart upload (1-based `part_number`); returns the ETag the
    /// caller must remember for `complete_multipart` — mirrors `UploadPart`. S3 requires
    /// every part except the last to be ≥ 5 MiB; the tus handler enforces that.
    pub async fn upload_part(
        &self,
        key: &str,
        upload_id: &str,
        part_number: i32,
        body: ByteStream,
        size: i64,
    ) -> Result<String> {
        let out = self
            .client
            .upload_part()
            .bucket(&self.bucket)
            .key(key)
            .upload_id(upload_id)
            .part_number(part_number)
            .body(body)
            .content_length(size)
            .send()
            .await
            .with_context(|| format!("s3 upload part {part_number}"))?;
        out.e_tag()
            .map(String::from)
            .with_context(|| format!("s3 upload part {part_number}: empty etag"))
    }

    /// Finalises the multipart upload, producing one object at `key` — mirrors
    /// `CompleteMultipart`. Parts must be in `part_number` order.
    pub async fn complete_multipart(
        &self,
        key: &str,
        upload_id: &str,
        parts: &[CompletedPart],
    ) -> Result<()> {
        let sdk_parts: Vec<S3CompletedPart> = parts
            .iter()
            .map(|p| {
                S3CompletedPart::builder()
                    .part_number(p.part_number)
                    .e_tag(&p.etag)
                    .build()
            })
            .collect();
        let completed = CompletedMultipartUpload::builder()
            .set_parts(Some(sdk_parts))
            .build();
        self.client
            .complete_multipart_upload()
            .bucket(&self.bucket)
            .key(key)
            .upload_id(upload_id)
            .multipart_upload(completed)
            .send()
            .await
            .context("s3 complete multipart")?;
        Ok(())
    }

    /// Discards a multipart upload (user cancel / stale-upload sweep) — mirrors
    /// `AbortMultipart`. Idempotent per the S3 spec.
    pub async fn abort_multipart(&self, key: &str, upload_id: &str) -> Result<()> {
        self.client
            .abort_multipart_upload()
            .bucket(&self.bucket)
            .key(key)
            .upload_id(upload_id)
            .send()
            .await
            .context("s3 abort multipart")?;
        Ok(())
    }
}
